feat: dynamic resizing of grid items

dan13ram opened the pull request:
## Overview

**What features/fixes does this PR include?**
This should fix the issue where the fixed height in `react-grid-layout` was raising problems when the content of the box was either too short or too long.

**Please provide the GitHub issue number**

Closes #1036 

## Follow up Improvement Ideas

- [ ] please list any improvement/ideas

## Implementation

**Describe technical (nontrivial / non-obvious) parts of your code**
It uses a ResizeObserver to observer for height changes and update the layout accordingly.

**Side effects**
<!---
Server deploy required?
Desktop apps?
MS Teams app?
Onboarding?
Backwards compatibility?
Performance sensitive changes?
Are translations required?
-->

## Assets

[Include screenshots/videos if it makes reviewing easier.]

vidvidvid reviewed (commented):
Lovely! I tested editing the layout and everything seems to be working fine! :)
I just found one tiny nuisance. When adding an empty block, the size of it is too large, but it gets set to the right size after refreshing the page:
![block](https://user-images.githubusercontent.com/35112939/150494945-8db4b15d-5f9d-485d-9942-ca15b2129293.gif)
 
Could this be solved as a part of this PR?

dan13ram commented:
Thank you @vidvidvid I'll take a look at that issue. I'll solve it within the PR

vidvidvid commented:
One more thing - there is a gap between the cover photo and the megamenu, idk how I didn't notice this earlier 😅 
<img width="1792" alt="Screenshot 2022-01-21 at 21 38 56" src="https://user-images.githubusercontent.com/35112939/150597192-c150c885-0c67-4160-b853-cb4cf572c1e2.png">


And while you're at it - if not reverting back to how the profile hero was positioned before the layout upgrade - might we consider removing or at least decreasing the padding on top? I think that the cover photo takes too much space now. @davort could we have your thoughts on this?
![padding_profile](https://user-images.githubusercontent.com/35112939/150596593-35363800-fe85-4f9b-afe5-4d7719f02a60.gif)

dan13ram commented:
@vidvidvid I've fixed that issue you mentioned and also removed the gap on top. I've kept the padding on top so that the cover photo remains visible.

vidvidvid commented:
But maybe we should wait until @dysbulic merges his pr so he won't have to do the rebasing yet again. @dysbulic what do you say?

vidvidvid commented:
Although it doesn't have any conflicts with `develop`, so there might be no conflicts when rebasing.

dan13ram commented:
I understand @vidvidvid We can wait for @dysbulic

dysbulic commented:
> But maybe we should wait until @dysbulic merges his pr so he won't have to do the rebasing yet again.

I'd sure appreciate it. I definitely want to see this one go through, but hopefully we can get a review done in the [meeting tomorrow](//discord.gg/Sy7jupey?event=933387177990168656) and get #943 merged.

dysbulic commented on packages/web/pages/player/[username].tsx:
```diff
@@ -178,37 +245,51 @@ export const Grid: React.FC<Props> = ({ player: initPlayer }): ReactElement => {
     setCurrentLayoutData,
   ] = useState<ProfileLayoutData>(savedLayoutData);
 
+  const itemsRef = useRef<(HTMLDivElement | null)[]>([]);
```
```suggestion
  const itemsRef = useRef<Array<HTMLElement>>([]);
```

dysbulic commented on packages/web/pages/player/[username].tsx:
```diff
@@ -178,37 +245,51 @@ export const Grid: React.FC<Props> = ({ player: initPlayer }): ReactElement => {
+    itemsRef.current = itemsRef.current.slice(0, currentLayoutItems.length);
+  }, [currentLayoutItems]);
+
+  const heights = useItemHeights(itemsRef, currentLayoutItems);
```
```suggestion
  const heights = useItemHeights(itemsRef.current, currentLayoutItems);
```

dysbulic commented on packages/web/pages/player/[username].tsx:
```diff
@@ -427,50 +499,47 @@ export const Grid: React.FC<Props> = ({ player: initPlayer }): ReactElement => {
+        {currentLayoutItems.map(({ boxKey, boxType, boxMetadata }, i) =>
           boxType === BoxType.PLAYER_ADD_BOX ? (
-            <Flex key={boxKey} className="gridItem">
+            <Flex key={boxKey} className="gridItem" id={boxKey}>
```
Why are you setting the `id`? Is it referenced by a stylesheet or something?

Also, the `Flex` tags for these two sections are the same, they can be moved to wrap the ternary and deduplicated.

dan13ram replied:
@dysbulic The `id` is needed for `getBoxKeyFromTarget` in line 166

dysbulic reviewed (commented):
Lots of little formatting quibbles, but the basic layout seems to make sense.

dan13ram commented:
@dysbulic Do take another look now if this is good to merge :)
